ArjunaJTS/jtax/classes/com/arjuna/ats/internal/jta/resources/jts/orbspecific/XAResourceRecord.java

```diff
@@ -909,14 +909,15 @@ public void forget() throws org.omg.CORBA.SystemException
             jtaxLogger.logger.trace("XAResourceRecord.forget for " + _tranID);
         }
 
-		handleForget() ;
+		if (!handleForget())
+		    throw new UNKNOWN(); // will cause log to be rewritten.
 
 		destroyState();
 
 		removeConnection();
 	}
 
-	private void handleForget()
+	private boolean handleForget()
 	{
 		if ((_theXAResource != null) && (_tranID != null))
 		{
```

Comment:
I think you need to move the _heuristic = TwoPhaseOutcome.FINISH_OK; to the same place as the local JTA one.